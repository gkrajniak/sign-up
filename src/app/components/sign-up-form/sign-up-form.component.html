<div class="card">
    <div class="card-body">
        <h5 class="card-title text-center">Sign Up</h5>
        <form [formGroup]="signupForm" (ngSubmit)="onSubmit()">
            <div class="mb-3">
                <label for="firstName" class="form-label">First Name<span class="text-danger">*</span></label>
                <input
                    type="text"
                    class="form-control"
                    id="firstName"
                    formControlName="firstName"
                    [class.is-invalid]="isFieldInvalid('firstName')" />
                <div class="invalid-feedback">First Name is required.</div>
            </div>
            <div class="mb-3">
                <label for="lastName" class="form-label">Last Name<span class="text-danger">*</span></label>
                <input
                    type="text"
                    class="form-control"
                    id="lastName"
                    formControlName="lastName"
                    [class.is-invalid]="isFieldInvalid('lastName')" />
                <div class="invalid-feedback">Last Name is required.</div>
            </div>
            <div class="mb-3">
                <label for="email" class="form-label">Email<span class="text-danger">*</span></label>
                <input type="email" class="form-control" id="email" formControlName="email" [class.is-invalid]="isFieldInvalid('email')" />
                <div class="invalid-feedback">Email is required and must be a valid email address.</div>
            </div>
            <div class="mb-3">
                <label for="password" class="form-label">Password<span class="text-danger">*</span></label>
                <input
                    type="password"
                    class="form-control"
                    id="password"
                    formControlName="password"
                    [class.is-invalid]="isFieldInvalid('password')" />
                <div class="invalid-feedback">Password is required.</div>
            </div>
            <div class="mb-3">
                <label for="confirmPassword" class="form-label">Confirm Password<span class="text-danger">*</span></label>
                <input
                    type="password"
                    class="form-control"
                    id="confirmPassword"
                    formControlName="confirmPassword"
                    [class.is-invalid]="isFieldInvalid('confirmPassword')" />
                <div class="invalid-feedback">Confirm Password is required.</div>
            </div>
            <div class="d-flex justify-content-end mt-4">
                <button type="submit" class="btn btn-primary" [disabled]="!signupForm.valid">Sign Up</button>
            </div>
        </form>
    </div>
</div>
