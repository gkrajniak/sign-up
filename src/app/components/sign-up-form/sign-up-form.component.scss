.display-block {
  display: block;
}

.toast-container {
  z-index: 10;
  width: fit-content;
  max-width: 300px;
  min-width: 200px;
  background-color: #f8f9fa;
  border: 1px solid #ddd;
  border-radius: 0.25rem;
  box-shadow: 0 0.125rem 0.25rem 0 rgba(0, 0, 0, 0.05);
}
