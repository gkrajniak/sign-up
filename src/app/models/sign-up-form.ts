import { FormControl } from '@angular/forms';

export interface SignUpForm {
    firstName: FormControl<string | null>;
    lastName: FormControl<string | null>;
    email: FormControl<string | null>;
    password: FormControl<string | null>;
}

export type FormField = keyof SignUpForm;
