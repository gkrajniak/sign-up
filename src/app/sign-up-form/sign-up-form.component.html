<div class="card">
    <div class="card-body">
        <h5 class="card-title text-center">Sign Up</h5>
        <form>
            <div class="mb-3">
                <label for="username" class="form-label">Username</label>
                <input type="text" class="form-control" id="username" placeholder="Enter your username" />
            </div>
            <div class="mb-3">
                <label for="password" class="form-label">Password</label>
                <input type="password" class="form-control" id="password" placeholder="Enter your password" />
            </div>
            <button type="submit" class="btn btn-primary">Sign In</button>
        </form>
    </div>
</div>
