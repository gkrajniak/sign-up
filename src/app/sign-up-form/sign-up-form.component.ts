import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';

@Component({
    selector: 'app-sign-up-form',
    standalone: true,
    imports: [CommonModule],
    templateUrl: './sign-up-form.component.html',
    styleUrl: './sign-up-form.component.scss',
})
export class SignUpFormComponent {}
