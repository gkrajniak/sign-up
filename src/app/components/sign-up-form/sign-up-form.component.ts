import { ChangeDetectionStrategy, Component, OnDestroy, OnInit } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormBuilder, FormGroup, ReactiveFormsModule, Validators } from '@angular/forms';
import { emailValidator } from '../../shared/validators/email.validator';
import { FormField, SignUpForm } from '../models/sign-up-form';
import { passwordValidator } from '../../shared/validators/password.validator';
import { Subject, takeUntil } from 'rxjs';

@Component({
    selector: 'app-sign-up-form',
    standalone: true,
    imports: [CommonModule, ReactiveFormsModule],
    templateUrl: './sign-up-form.component.html',
    styleUrl: './sign-up-form.component.scss',
    changeDetection: ChangeDetectionStrategy.OnPush,
})
export class SignUpFormComponent implements OnInit, OnDestroy {
    signupForm!: FormGroup<SignUpForm>;
    private destroy$: Subject<void> = new Subject<void>();
    protected showPassword = false;

    constructor(private fb: FormBuilder) {}

    ngOnInit(): void {
        this.signupForm = this.fb.group({
            firstName: ['', Validators.required],
            lastName: ['', Validators.required],
            email: ['', [Validators.required, emailValidator()]],
            password: ['', Validators.required],
        });

        this.signupForm.controls.password.addValidators(passwordValidator(this.signupForm));

        this.signupForm.controls.firstName.valueChanges.pipe(takeUntil(this.destroy$)).subscribe((v) => {
            this.signupForm.controls.password.updateValueAndValidity();
        });
        this.signupForm.controls.lastName.valueChanges.pipe(takeUntil(this.destroy$)).subscribe((v) => {
            this.signupForm.controls.password.updateValueAndValidity();
        });
    }

    ngOnDestroy() {
        this.destroy$.next();
        this.destroy$.complete();
    }

    onSubmit() {
        if (this.signupForm.valid) {
            console.log('Form submitted:', this.signupForm.value);
        }
    }

    protected isFieldInvalid(fieldName: FormField): boolean {
        const control = this.signupForm.get(fieldName);
        return control ? control.invalid && control?.touched : false;
    }

    protected getPasswordErrorMsg() {
        const ctrl = this.signupForm.controls.password;
        if (ctrl.hasError('required')) {
            return 'Password is required.';
        }

        const invalidPassword = ctrl.getError('invalidPassword');
        if (invalidPassword) {
            return invalidPassword.message;
        }
    }

    protected togglePasswordVisibility() {
        this.showPassword = !this.showPassword;
    }

    protected formNotValid() {
        return !this.signupForm.valid;
    }
}
