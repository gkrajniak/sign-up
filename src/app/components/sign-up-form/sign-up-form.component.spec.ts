import { ComponentFixture, TestBed } from '@angular/core/testing';

import { SignUpFormComponent } from './sign-up-form.component';
import { ChangeDetectorRef } from '@angular/core';

describe(SignUpFormComponent, () => {
    let component: SignUpFormComponent;
    let fixture: ComponentFixture<SignUpFormComponent>;
    let changeDetectorRef: ChangeDetectorRef;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
            imports: [SignUpFormComponent],
        }).compileComponents();

        fixture = TestBed.createComponent(SignUpFormComponent);
        component = fixture.componentInstance;
        changeDetectorRef = fixture.componentRef.injector.get(ChangeDetectorRef);
        fixture.detectChanges();
    });

    it('should create component', () => {
        expect(component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(fixture).toMatchSnapshot();
    });

    describe('check ui state', () => {
        it('should have submit button disabled', () => {
            // given
            const formGroup = component.signupForm;

            // when
            changeDetectorRef.detectChanges();

            // then
            expect(formGroup.valid).toBeFalsy();
            const submitButton = fixture.nativeElement.querySelector('button[type="submit"]');
            expect(submitButton.disabled).toBeTruthy();
        });

        it('should enable the submit button when the form is valid', () => {
            // given
            const formGroup = component.signupForm;
            formGroup.controls.firstName.setValue('John');
            formGroup.controls.lastName.setValue('Doe');
            formGroup.controls.email.setValue('johndoe@example.com');
            formGroup.controls.password.setValue('StrongPassword123');

            // when
            changeDetectorRef.detectChanges();

            // then
            expect(formGroup.valid).toBeTruthy();
            const submitButton = fixture.nativeElement.querySelector('button[type="submit"]');
            expect(submitButton.disabled).toBeFalsy();
        });

        it('should enable the submit button only when all required form controls are valid', () => {
            const formGroup = component.signupForm;

            // Set first name
            formGroup.controls.firstName.setValue('John');
            changeDetectorRef.detectChanges();

            expect(formGroup.valid).toBeFalsy();
            let submitButton = fixture.nativeElement.querySelector('button[type="submit"]');
            expect(submitButton.disabled).toBeTruthy();

            // Set last name
            formGroup.controls.lastName.setValue('Doe');
            changeDetectorRef.detectChanges();

            expect(formGroup.valid).toBeFalsy();
            submitButton = fixture.nativeElement.querySelector('button[type="submit"]');
            expect(submitButton.disabled).toBeTruthy();

            // Set email
            formGroup.controls.email.setValue('johndoe@example.com');
            changeDetectorRef.detectChanges();

            expect(formGroup.valid).toBeFalsy();
            submitButton = fixture.nativeElement.querySelector('button[type="submit"]');
            expect(submitButton.disabled).toBeTruthy();

            // Set password
            formGroup.controls.password.setValue('StrongPassword123');
            changeDetectorRef.detectChanges();

            expect(formGroup.valid).toBeTruthy();
            submitButton = fixture.nativeElement.querySelector('button[type="submit"]');
            expect(submitButton.disabled).toBeFalsy();
        });
    });

    describe('check validation', () => {
        it('should not execute submit button', function () {
            // when
            component.onSubmit();

            // then todo gkr
        });
    });
});
