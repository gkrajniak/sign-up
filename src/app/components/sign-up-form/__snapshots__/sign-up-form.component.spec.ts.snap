// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpFormComponent should match snapshot 1`] = `
<app-sign-up-form
  changeDetectorRef={[Function ViewRef$1]}
  destroy$={[Function Subject]}
  fb={[Function _FormBuilder]}
  showPassword="false"
  signUpService={[Function Object]}
  signupForm={[Function FormGroup]}
  singUpResult={[Function Object]}
>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title text-center"
      >
        Sign Up
      </h5>
      <form
        class="ng-untouched ng-pristine ng-invalid"
        novalidate=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="firstName"
          >
            First Name
            <span
              class="text-danger"
            >
              *
            </span>
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            data-test="firstName-input"
            formcontrolname="firstName"
            id="firstName"
            type="text"
          />
          <div
            class="invalid-feedback"
            data-test="invalid-feedback-firstName"
          >
            First Name is required.
          </div>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="lastName"
          >
            Last Name
            <span
              class="text-danger"
            >
              *
            </span>
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            data-test="lastName-input"
            formcontrolname="lastName"
            id="lastName"
            type="text"
          />
          <div
            class="invalid-feedback"
            data-test="invalid-feedback-lastName"
          >
            Last Name is required.
          </div>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="email"
          >
            Email
            <span
              class="text-danger"
            >
              *
            </span>
          </label>
          <input
            class="form-control ng-untouched ng-pristine ng-invalid"
            data-test="email-input"
            formcontrolname="email"
            id="email"
            type="email"
          />
          <div
            class="invalid-feedback"
            data-test="invalid-feedback-email"
          >
            Email is required and must be a valid email address.
          </div>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
            <span
              class="text-danger"
            >
              *
            </span>
          </label>
          <div
            class="input-group"
          >
            <input
              class="form-control ng-untouched ng-pristine ng-invalid"
              data-test="password-input"
              formcontrolname="password"
              id="password"
              type="password"
            />
            <button
              class="btn btn-outline-secondary"
              data-test="togglePassword-button"
              type="button"
            >
              <i
                class="bi bi-eye"
              />
            </button>
          </div>
        </div>
        <div
          class="d-flex justify-content-end mt-4"
        >
          <button
            class="btn btn-primary"
            data-test="submit-button"
            disabled=""
            type="submit"
          >
            Sign Up
          </button>
        </div>
      </form>
    </div>
  </div>
</app-sign-up-form>
`;
